replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/AbstractPersistable.java

```diff
@@ -0,0 +1,178 @@
+ * to manipulate that data.
+ *
+ * @param <P> the type of pojo associated with this persistable
+ */
+public abstract class AbstractPersistable<P extends Pojo> implements Persistable {
+  /** A string representing the type of this object. Used when generating exception or logs. */
+  protected final String type;
+
+  private String id;
+
+  protected AbstractPersistable(String type) {
+    this.type = type;
+    this.id = UUID.randomUUID().toString();
+  }
+
+  protected AbstractPersistable(String type, @Nullable String id) {
+    this.type = type;
+    this.id = id;
+  }
+
+  @Override
+  public String getId() {
+    return id;
+  }
+
+  /**
+   * Serializes this object to the specified pojo that can be serialized to a spring-data supported
+   * data store. Any implementation of this method in a subclass should first make a call to the
+   * super version of this method before serializing its own properties.
+   *
+   * @param pojo the pojo to serialize this object to
+   * @return <code>pojo</code> for chaining
+   * @throws ReplicationPersistenceException if an error occurs while trying to serialize this
+   *     object
+   */
+  protected P writeTo(P pojo) {
+    pojo.setId(id);
+    return pojo;
+  }
+
+  /**
+   * Deserializes the specified pojo into this persistable object. This method is responsible for
+   * performing any required conversion to the latest version. Any implementation of this method in
+   * a subclass should first make a call to the super version before performing its own
+   * functionality.
+   *
+   * @param pojo the pojo to deserialize
+   * @throws ReplicationPersistenceException if an error occurs while trying to deserialize the pojo
+   *     object
+   */
+  protected void readFrom(P pojo) {
+    setOrFailIfNull("identifier", pojo::getId, this::setId);
+  }
+
+  /**
+   * Useful method that can be used to validate if the value of a field is not <code>null</code> in
+   * the pojo before setting it in this object. If it is <code>null</code> than an exception is
+   * thrown otherwise the object's corresponding field would be set accordingly. The pojo field's
+   * value is retrieved using a consumer (e.g. <code>pojo::getName</code>) and set in this object
+   * using a supplier (e.g. <code>this::setName</code>).
+   *
+   * @param field the name of the field being check for presence (used in the exception message)
+   * @param supplier a supplier capable of retrieving the pojo's current value for the field
+   * @param consumer a consumer capable of updating this object with the field's value if it is
+   *     defined
+   * @param <F> the field's type
+   * @throws ReplicationPersistenceException if the field's value as supplied by <code>supplier
+   *     </code> is <code>null</code>
+   */
+  protected <F> void setOrFailIfNull(
+      String field, Supplier<? extends F> supplier, Consumer<? super F> consumer) {
```

Comment:
I was wondering the same thing

Reply:
Really good question. I guess I started with the need for a consumer and decided to keep it consistent. It also allows me to potentially control potential exceptions in the future while retrieving (although currently I don't need to do that).  Since this method is intended to validate the value before setting it, I saw no reasons for allowing you to pass a constant since you wouldn't need to validate it and as such, could have set it directly without the need for this method.

I am good either way.